src/fetch/rewards.py

```diff
@@ -0,0 +1,167 @@
+    primary_rewards_df = compute_primary_rewards(batch_data_df, converter)
+
+    # secondary rewards
+    secondary_rewards_df = compute_secondary_rewards_df(
+        batch_data_df, primary_rewards_df, converter
+    )
+
+    # quote rewards
+    quote_rewards_df = compute_quote_rewards(trade_data_df, converter)
+
+    # merge different rewards
+    solver_rewards_df = (
+        primary_rewards_df.merge(secondary_rewards_df, on="solver", how="outer")
+    ).merge(quote_rewards_df, on="solver", how="outer")
+
+    return solver_rewards_df
+
+
+def compute_primary_rewards(
+    batch_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute primary rewards from batch data"""
+    primary_rewards_df = (
+        batch_data_df.apply(compute_primary_reward_datum, axis=1, result_type="expand")
+        .rename(columns={0: "solver", 1: "primary_reward_eth"})
+        .astype(object)
+        .groupby("solver")
+        .sum()
+    )
+    primary_rewards_df["primary_reward_cow"] = (
+        primary_rewards_df["primary_reward_eth"]
+        .apply(converter.eth_to_token)
+        .astype(object)
+    )
+    return primary_rewards_df
+
+
+def compute_primary_reward_datum(row: Series) -> tuple[str, int]:
+    """Compute the reward of an auction from the winning and references scores"""
+    solver = row["winning_solver"]
+    if row["tx_hash"] is None:
+        reward = -min(row["reference_score"], EPSILON_LOWER)
+    else:
+        reward = min(row["winning_score"] - row["reference_score"], EPSILON_UPPER)
+
+    return solver, reward
+
+
+def compute_secondary_rewards_df(
+    batch_data_df: DataFrame, primary_rewards_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute secondary rewards from batch data and primary rewards"""
+    secondary_allocation = max(
+        min(
+            PERIOD_BUDGET_COW - primary_rewards_df["primary_reward_cow"].sum(),
+            converter.eth_to_token(CONSISTENCY_REWARD_CAP_ETH),
+        ),
+        0,
+    )
+    secondary_rewards_df = (
+        batch_data_df["participating_solvers"]
+        .explode()
+        .value_counts()
+        .reset_index()
+        .rename(columns={"participating_solvers": "solver", "count": "num_auctions"})
+        .astype(object)
+    )
+    participation_total = secondary_rewards_df["num_auctions"].sum()
+    secondary_rewards_df["secondary_reward_cow"] = (
+        secondary_allocation * secondary_rewards_df["num_auctions"]
+    ) // participation_total
+    secondary_rewards_df["secondary_reward_eth"] = (
+        secondary_rewards_df["secondary_reward_cow"]
+        .apply(converter.token_to_eth)
+        .astype(object)
+    )
+    secondary_rewards_df = secondary_rewards_df.drop("num_auctions", axis=1)
+
+    return secondary_rewards_df
+
+
+def compute_quote_rewards(
+    trade_data_df: DataFrame, converter: TokenConversion
+) -> DataFrame:
+    """Compute quote rewards from batch data"""
+    quote_rewards_df = (
```

Comment:
as pythonic as it can get, or at least this is how it looks to me.....

Reply:
It is not super pythonic. The function which is applied to every row should probably be its own function. The rest is then working around subtleties of pandas.